File: 409-longest-palindrome/longest-palindrome.java
class Solution {
    public int longestPalindrome(String s) {
       HashMap<Character, Integer> freq = new HashMap<>(); // Array hataya, map banaya
        
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            freq.put(c, freq.getOrDefault(c, 0) + 1); // Pass 1: Dynamic entry
        }
        
        int length = 0;
        boolean hasOdd = false;
        
        // Map ki saare values (counts) par loop chalaya
        for (int count : freq.values()) {
            length += (count / 2) * 2;
            
            if (count % 2 != 0) {
                hasOdd = true;
            }
        }
        
        if (hasOdd) length += 1;
        return length;

        // // by vector
        //  int[] freq = new int[128]; // Correct syntax [128] ke sath fixed array
        
        // for (int i = 0; i < s.length(); i++) {
        //     freq[s.charAt(i)]++; // Pass 1: Frequency calculation
        // }
        
        // int length = 0;
        // boolean hasOdd = false;
        
        // for (int count : freq) {
        //     length += (count / 2) * 2;
            
        //     if (count % 2 != 0) {
        //         hasOdd = true;
        //     }
        // }
        
        // if (hasOdd) length += 1;
        // return length;
    }
}

File: 409-longest-palindrome/longest-palindrome.cpp
class Solution {
public:
    int longestPalindrome(string s) {
         // by vector freq method 
        vector<int> freq(128, 0);
        // for(i = 0;i < s.length();i++){
            // char c = s[i];
        for(char c : s){
            freq[c]++;
        }
        int length = 0;
        bool has_odd = false;
        for(int i = 0;i < 128;i++){
            int count = freq[i];
            length += (count / 2) * 2 ;//integer dision trick

            if(count % 2 != 0){
                has_odd = true;
            }
        }
        if(has_odd){
            length += 1;
        }
        return length;
    }
};